src/native/libs/System.Native/pal_interfaceaddresses.c

```diff
@@ -95,14 +100,74 @@ static inline uint8_t mask2prefix(uint8_t* mask, int length)
 
     return len;
 }
+
+#if !HAVE_IFADDRS && TARGET_ANDROID
+// This structure is exactly the same as struct ifaddrs defined in ifaddrs.h but since the header
+// might not be available (e.g., in bionics used in Android before API 24) we need to mirror it here
+// so that we can dynamically load the getifaddrs function and use it.
+struct ifaddrs
+{
+	struct ifaddrs *ifa_next;
+	char *ifa_name;
+	unsigned int ifa_flags;
+	struct sockaddr *ifa_addr;
+	struct sockaddr *ifa_netmask;
+	union
+	{
+		struct sockaddr *ifu_broadaddr;
+		struct sockaddr *ifu_dstaddr;
+	} ifa_ifu;
+	void *ifa_data;
+};
+#endif
+
+#if !HAVE_GETIFADDRS && TARGET_ANDROID
+// Try to load the getifaddrs and freeifaddrs functions manually.
+// This workaround is necessary on Android prior to API 24 and it can be removed once
+// we drop support for earlier Android versions.
+static int (*getifaddrs)(struct ifaddrs**) = NULL;
+static void (*freeifaddrs)(struct ifaddrs*) = NULL;
+
+static bool ensure_getifaddrs_is_loaded()
+{
+    static atomic_bool initialized = false;
+    static atomic_bool access_guard = true;
+
+    while (!atomic_load(&initialized))
```

Comment:
Is `pthread_once` available on Android? It would be better to use here than the busy waiting loop.

If two threads happen to try to initialize this at the same time, one of them may end up busy waiting for multiple miliseconds here in this implementation that's not efficient. `pthread_once` does not have this problem.

Reply:
`pthread_once` is available and it also uses `atomic_load_explicit` in a loop + futex syscalls, so it will most likely be heavier than this simple loop. The likelihood of two threads hitting this code at exactly the same time is so low that, IMO, it's not worth worrying about.

Reply:
Yes, `pthread_once` has a loop similar to this one, except that it also handles the case that I have pointed out. The extra cost of `pthread_once` is negligible given what this whole method does.

The #1 rule for lock-free programming: Do not write lock-free algorithm yourselves unless you absolutely have to.